copy/copy.go

```diff
@@ -282,26 +354,127 @@ func (c *copier) copy(ctx context.Context, src, target string, overwriteTargetMe
 	return nil
 }
 
-func (c *copier) copyDirectory(ctx context.Context, src, dst string, stat os.FileInfo, overwriteTargetMetadata bool) (bool, error) {
+func (c *copier) include(path string, fi os.FileInfo) (bool, bool, error) {
+	matched := false
+	partial := true
+	for _, pattern := range c.includePatterns {
+		if fi.IsDir() {
+			pattern = strings.TrimSuffix(pattern, string(filepath.Separator))
+		}
+
+		if ok, p := prefix.Match(pattern, path); ok {
+			matched = true
+			if !p {
+				partial = false
+				break
+			}
+		}
+	}
+
+	if !matched {
+		return false, false, nil
+	}
+	if fi.IsDir() {
+		return true, !partial, nil
+	}
+	return !partial, !partial, nil
+}
+
+func (c *copier) exclude(path string, fi os.FileInfo) (bool, error) {
+	if c.excludePatternMatcher == nil {
+		return false, nil
+	}
+
+	m, err := c.excludePatternMatcher.Matches(path)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to match excludepatterns")
+	}
+	if m {
+		if fi.IsDir() && c.excludePatternMatcher.Exclusions() {
```

Comment:
Not for this PR but this looks non-optimal. I guess `!` is rarely used in practice.

Reply:
Modeled on this code from `walker.go`: https://github.com/tonistiigi/fsutil/blob/8599091d69851a37058eec95da25c794ac060a0a/walker.go#L131